### ISSUE TITLE
test_duplicator.py generates a wrong most right column

### ISSUE THREAD

**Issue by vicente-gonzalez-ruiz** [Sistemas-Multimedia/MCDWT#98]
Could this issue be related to closed issue #85?

**Comment by juanjonrg** [comment]
Yes, it generates black pixels when mapping pixels out of bounds. What should it do in those cases?

**Comment by vicente-gonzalez-ruiz** [comment]
Why the Motion Estimator selects pixels that are out of the reference images? Obviously, in a camera panning, at least one of the reference images (the "youngest") has valid pixels. In that case, at most, the predicted pixels should have half of the intensity than the rest of pixels in the prediction image.

@juanjonrg , do you think that if the movement in the sequence were different, this problem should happen in a different area? Or this is a problem of the optical flow estimation algorithm provided by OpenCV?

**Comment by juanjonrg** [comment]
Hi Vicente,

Then there's definitely a bug in the motion compensation. I understood it as:
1. Compute the optical flow between two frames A and C
2. Apply that optical flow to a frame B, using only pixels from B.

If the camera is moving to the right, the optical flow of (almost) all the pixels will point to the right and they will be replaced with pixels to the right of them. This is why the rightmost column gets, in this case, zeroed.

Is this behaviour wrong? What should we do in those cases?

**Comment by vicente-gonzalez-ruiz** [comment]
@juanjonrg , the interpolator inputs 2 images A and C and outputs 3, A, B=(A'+C')/2 and C, where A'=MC(A, A->B) and C'=MC(C, C->B), MC = motion compensation, and "X->Y" denotes the motion to approximate Y using X. So, if you implemented what you have said, yes, there is a misunderstanding.

However, I don't understand how do you read 5 images and write 9, doing what you say.

Thanks!

**Comment by vicente-gonzalez-ruiz** [comment]
Hi @juanjonrg . Please, could you modify your code to create the interpolated images in the oposite direction and finally generate the interpolated images as the average between the forward and the backward prediction? In other words, if I'm not wrong, at this moment you have computed a interpolated frame B as MC(A, ME(A,C)/2)) where MC(X,Y) is the interpolated image generated by moving the pixels of X using the motion field Y, which is the optical flow between the adjacent images of X, divided by 2. My proposal would be:

I =  ( MC(A, ME(A,C)/2) + MC(C, ME(C,A)/2) )/2

Thnx!

**Comment by juanjonrg** [comment]
Hi @vicente-gonzalez-ruiz,

The duplicator outputs 2n - 1 frames from n frames, as it generates an estimated frame for each pair of input frames. It also outputs the (unmodified) input frames in the right order.

I have tried implementing `I = ( MC(A, ME(A,C)/2) + MC(C, ME(C,A)/2) )/2` but the black columns are still noticeable. However, if we only do `I = MC(C, ME(C, A)/2)` it shows a pretty good result without the black columns. 

I will send a pull request with that change for you to review.

**Comment by vicente-gonzalez-ruiz** [comment]
`I have tried implementing I = ( MC(A, ME(A,C)/2) + MC(C, ME(C,A)/2) )/2 but the black columns are still noticeable. However, if we only do I = MC(C, ME(C, A)/2) it shows a pretty good result without the black columns.` -> I think that this is because the camera panning is in the right direction.

Yes, the right column is as it is (wrong) due to the algorithm we have used. In fact, to solve this we should control the prediction of each pixel taking into account the number of available pixel references, which in some situations could be even zero (think of an object that only appears in one image).
